Fix external ABI type name for contracts

Marenz opened the pull request:
I hope this is the right way to fix this ..

fixes #6435

Marenz committed:
Fix external ABI type name for contracts

chriseth commented:
Hm, I'm not sure this is the correct fix. `ContractType::interfaceType` correctly returns address in the non-library case. Perhaps the flag for that function is not set correctly when it is called?

Marenz commented:
As far as I understood the code, the name is generated from https://github.com/ethereum/solidity/blob/develop/libsolidity/ast/Types.cpp#L3148
I don't see where the function `ContractType::interfaceType` is used for this

ekpyron commented:
@Marenz To see where ``ContractType::interfaceType`` is used: In the line you're linking ``_t`` stems from ``parameterTypes`` which is the parameter types of the ``interfaceFunctionType()`` of the function type, which is the ``ìnterfaceType``s of the function's paramter... however ``interfaceFunctionType()`` doesn't get the local ``inLibrary`` in ``externalSignature`` passed down, so it calls ``interfaceType(true)`` instead of ``interfaceType(false)``... does that help?

ekpyron commented on libsolidity/ast/Types.cpp:
```diff
@@ -3139,13 +3139,13 @@ string FunctionType::externalSignature() const
 	{
-		solAssert(_t, "Parameter should have external type.");
-		string typeName = _t->signatureInExternalFunction(inLibrary);
+		auto const& _it = _t->interfaceType(inLibrary);
```
```suggestion
		auto const& it = _t->interfaceType(inLibrary);
```

ekpyron commented on libsolidity/ast/Types.cpp:
```diff
@@ -3139,13 +3139,13 @@ string FunctionType::externalSignature() const
-		solAssert(_t, "Parameter should have external type.");
-		string typeName = _t->signatureInExternalFunction(inLibrary);
+		auto const& _it = _t->interfaceType(inLibrary);
+		solAssert(_it.get(), "Parameter should have external type.");
```
```suggestion
		solAssert(!!it, "Parameter should have external type.");
```
(not sure about the ``!!`` myself, but it's used a lot...)

Marenz replied:
that doesn't actually compile. I'd need to use `!!it.get()`

ekpyron commented on libsolidity/ast/Types.cpp:
```diff
@@ -3139,13 +3139,13 @@ string FunctionType::externalSignature() const
+		auto const& _it = _t->interfaceType(inLibrary);
+		solAssert(_it.get(), "Parameter should have external type.");
+
+		string typeName = _it.get()->signatureInExternalFunction(inLibrary);
```
```suggestion
		string typeName = it->signatureInExternalFunction(inLibrary);
```

ekpyron reviewed (commented):
Some style remarks.
And I wonder whether we should add a comment to ``interfaceFunctionType()`` where the parameters are transformed using ``interfaceType`` stating that if one edit's that one likely has to edit ``externalSignature`` as well, just to be sure.

ekpyron commented:
Thinking about it: maybe we should extract the transformation of the parameter types via ``interfaceType(...)`` into a function (like ``interfaceParameterTypes(bool)``) that can be used from ``externalSignature`` and from ``interfaceFunctionType`` alike? That'd ensure that nobody changes one without the other...

Marenz commented:
Sounds good to me

Marenz commented:
Updated according to suggestions

Marenz commented:
Note: I used `std::move()` first, but had to remove it again because

```
/root/project/libsolidity/ast/Types.cpp:162:9: error: moving a local object in a return statement prevents copy elision [-Werror,-Wpessimizing-move]
        return std::move(transformed);
               ^
/root/project/libsolidity/ast/Types.cpp:162:9: note: remove std::move call here
        return std::move(transformed);
```

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -147,6 +147,21 @@ bool fitsIntoBits(bigint const& _value, unsigned _bits, bool _signed)
+{
+	TypePointers transformed;
+
+	for (auto type: _parameters)
```
```suggestion
	for (auto const& type: _parameters)
```

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -147,6 +147,21 @@ bool fitsIntoBits(bigint const& _value, unsigned _bits, bool _signed)
+
+	for (auto type: _parameters)
+	{
+		if (auto ext = type->interfaceType(_inLibrary).get())
```
```suggestion
		if (TypePointer ext = type->interfaceType(_inLibrary).get())
```

otherwise, the `.get()` looks like the `.get` function on shared_ptr and thus it might seem like `ext` is a raw pointer.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -147,6 +147,21 @@ bool fitsIntoBits(bigint const& _value, unsigned _bits, bool _signed)
+		if (auto ext = type->interfaceType(_inLibrary).get())
+			transformed.push_back(ext);
+		else
+			return {};
```
Callers of this function cannot easily distinguish between an error case and an invocation on the empty vector. Why not use a `Result<TypePointers>` and also hand over the error message from the `interfaceType` call?

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -2906,23 +2921,18 @@ FunctionTypePointer FunctionType::interfaceFunctionType() const
-	TypePointers paramTypes;
-	TypePointers retParamTypes;
+	TypePointers paramTypes =
+		transformParametersToExternal(m_parameterTypes, isLibraryFunction);;
```
Double `;`

Marenz commented:
Update to fix comments

chriseth commented:
Needs a changelog entry.

chriseth commented on test/libsolidity/SolidityEndToEndTest.cpp:
```diff
@@ -15607,6 +15607,38 @@ BOOST_AUTO_TEST_CASE(contract_name)
+			}
+		}
+	)";
+	compileAndRun(sourceCode, 0, "Another");
```
Why do you need to compile this?

Marenz replied:
I tried to copy the pattern I saw in the other tests

chriseth commented on test/libsolidity/SolidityEndToEndTest.cpp:
```diff
@@ -15607,6 +15607,38 @@ BOOST_AUTO_TEST_CASE(contract_name)
+		}
+	)";
+	compileAndRun(sourceCode, 0, "Another");
+	compileAndRun(sourceCode, 0, "ClientReceipt", bytes(), map<string, Address>{{"Another", m_contractAddress}});
```
`Another` is not a library, so you do not need to provide linking information here.

chriseth commented on test/libsolidity/SolidityEndToEndTest.cpp:
```diff
@@ -15607,6 +15607,38 @@ BOOST_AUTO_TEST_CASE(contract_name)
+		library ClientReceipt {
+			event Deposit(Another indexed _from, bytes32 indexed _id, uint _value);
+			function deposit(bytes32 _id) public {
+				Another a;
```
You could just re-use `Test` here.

chriseth commented on libsolidity/ast/Types.cpp:
```diff
@@ -3139,13 +3149,15 @@ string FunctionType::externalSignature() const
 
 	// "inLibrary" is only relevant if this is not an event.
 	bool const inLibrary = kind() != Kind::Event && dynamic_cast<ContractDefinition const&>(*m_declaration->scope()).isLibrary();
-	FunctionTypePointer external = interfaceFunctionType();
```
If `interfaceFunctionType` is not called anymore, we ignore the return parameters. Is that intended?

chriseth replied:
Perhaps we should also change the implementation of  `interfaceFunctionType` to include the special case about events in the line above.

Marenz replied:
Originally, the return value of `interfaceFunctionType()` was only used to get the `parameterTypes()`, so we basically did extra work for something we didn't even use eventually..

Regarding the special case: That was my original approach in this PR, @ekpyron suggested a different approach here: https://github.com/ethereum/solidity/pull/6473#issuecomment-480871065

Our fear was that it might have unforeseen effects on other functions using `interfaceFunctionType()`

Marenz commented:
Updated

chriseth commented:
I think we should still change `interfaceFunctionType` to make it consistent. Also: Please add a buglist entry.

chriseth commented:
I just checked - changing `interfaceFunctionType` should not have unwanted effects.

Marenz commented:
Updated

chriseth commented on docs/bugs.json:
```diff
@@ -1,4 +1,12 @@
+        "description": "Instead of using the type 'address' for hashing, the actual contract name was used, leading to a wrong hash in the ABI.",
+        "introduced": "0.3.0",
+        "fixed": "0.5.8",
+        "severity": "low"
```
I would say it is "very low".

chriseth commented on docs/bugs.json:
```diff
@@ -1,4 +1,12 @@
 [
+    {
+        "name": "IncorrectEventABIInLibraries",
+        "summary": "Contract types used in events in libraries cause an incorrect hash in the ABI",
```
```suggestion
        "summary": "Contract types used in events in libraries cause an incorrect event signature hash",
```

chriseth commented on docs/bugs.json:
```diff
@@ -1,4 +1,12 @@
+    {
+        "name": "IncorrectEventABIInLibraries",
+        "summary": "Contract types used in events in libraries cause an incorrect hash in the ABI",
+        "description": "Instead of using the type 'address' for hashing, the actual contract name was used, leading to a wrong hash in the ABI.",
```
```suggestion
        "description": "Instead of using the type `address` in the hashed signature, the actual contract name was used, leading to a wrong hash in the logs.",
```

chriseth commented on Changelog.md:
```diff
@@ -24,6 +24,7 @@ Compiler Features:
 Bugfixes:
  * SMTChecker: Implement Boolean short-circuiting.
  * SMTChecker: SSA control-flow did not take into account state variables that were modified inside inlined functions that were called inside branches.
+ * Type System: Use correct type name for contracts in event parameters
```
... used in libaries.

Perhaps also add: This affected code generation.

Marenz commented:
Fixed comments

chriseth commented on docs/bugs.json:
```diff
@@ -1,4 +1,12 @@
 [
+    {
+        "name": "IncorrectEventABIInLibraries",
```
Ah actually this also used "ABI" which I don't think is very relevant here. Perhaps `IncorrectEventSignatureInLibraries`?

Marenz commented:
Renamed bug

chriseth commented:
I think you have to update the bug list, at least that is what the mac test fails on.

chriseth commented:
Perhaps it also just ran on outdated code...

Marenz commented:
the problem was, that I modified the end-of-file newline and the script reverted that change and thus the file was not considered up-to-date. Fixed it now
